sherpa-onnx/csrc/offline-funasr-nano-model.cc

```diff
@@ -0,0 +1,1037 @@
+  size_t n = 1;
+  for (auto d : shape) {
+    if (d <= 0) return 0;
+    n *= static_cast<size_t>(d);
+  }
+  return n;
+}
+
+static inline bool TensorIsCpu(const Ort::Value &v) {
+  if (!v.IsTensor()) return true;
+  auto mi = v.GetTensorMemoryInfo();
+  return mi.GetDeviceType() == OrtMemoryInfoDeviceType_CPU;
+}
+
+static inline void AssertTensorIsCpu(const Ort::Value &v, const char *what) {
+  if (!v.IsTensor()) return;
+  auto mi = v.GetTensorMemoryInfo();
+  if (mi.GetDeviceType() != OrtMemoryInfoDeviceType_CPU) {
+    SHERPA_ONNX_LOGE("%s: expected CPU tensor but got device_type=%d device_id=%d",
+                     what, (int)mi.GetDeviceType(), mi.GetDeviceId());
+    SHERPA_ONNX_EXIT(-1);
+  }
+}
+
+static inline std::string ToLower(std::string s) {
+  std::transform(s.begin(), s.end(), s.begin(),
+                 [](unsigned char c) -> char {
+                   return static_cast<char>(std::tolower(c));
+                 });
+  return s;
+}
+
+static inline bool IsCudaProvider(const std::string &provider) {
+  auto p = ToLower(provider);
+  // Keep it conservative. We only enable IO binding policy below when we
+  // are on CUDA; other EPs keep the existing behavior.
+  return p == "cuda" || (p.size() > 4 && p.rfind("cuda", 0) == 0);
+}
+
+// Get the element type of a session input tensor.
+static inline ONNXTensorElementDataType GetSessionInputElemType(
+    Ort::Session *sess, size_t input_index) {
+  auto ti = sess->GetInputTypeInfo(input_index);
+  auto t  = ti.GetTensorTypeAndShapeInfo();
+  return static_cast<ONNXTensorElementDataType>(t.GetElementType());
+}
+
+// Get the element type of a session output tensor.
+static inline ONNXTensorElementDataType GetSessionOutputElemType(
+    Ort::Session *sess, size_t output_index) {
+  auto ti = sess->GetOutputTypeInfo(output_index);
+  auto t  = ti.GetTensorTypeAndShapeInfo();
+  return static_cast<ONNXTensorElementDataType>(t.GetElementType());
+}
+
+template <typename T>
+static Ort::Value AllocTensor(OrtAllocator *alloc,
+                              const std::vector<int64_t> &shape) {
+  return Ort::Value::CreateTensor<T>(alloc, shape.data(), shape.size());
+}
+
+template <>
+Ort::Value AllocTensor<uint16_t>(OrtAllocator *alloc,
+                                 const std::vector<int64_t> &shape) {
+  return Ort::Value::CreateTensor(alloc, shape.data(), shape.size(),
+                                  ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT16);
+}
+
+// Convert tensor to float32, handling both float16 and float32 inputs.
+// NOTE: This helper assumes the input tensor is on CPU memory.
+// The caller must ensure the tensor is on CPU (e.g., via IO Binding).
+static Ort::Value CastToFloat32(Ort::Value in, OrtAllocator *alloc) {
+  if (!in.IsTensor()) return in;
+  auto info = in.GetTensorTypeAndShapeInfo();
+  auto shape = info.GetShape();
+  size_t n = NumelFromShape(shape);
+  if (n == 0) return in;
+  auto et = static_cast<ONNXTensorElementDataType>(info.GetElementType());
```

Comment:
```suggestion
  auto et = info.GetElementType();
```

The return type is already `ONNXTensorElementDataType`, so there is no need to call `static_cast<>`.